File: Coding_Questions/Python/MaxSubsetSumNoAdjacent.py
# ---Max Subset Sum No Adjacent---

from typing import List


def max_subset_sum_no_adjacent(array: List[int]) -> int:
    if not len(array):
        return 0

    for i in range(len(array)):
        if i == 0 or i == 1:
            continue
        array[i] += max(array[:i-1])

    return max(array)


print(max_subset_sum_no_adjacent([75, 105, 120, 75, 90, 135]))  # 330

File: Coding_Questions/C++/MaxSubsetSumNoAdjacent.cpp
// ---Max Subset Sum No Adjacent---

#include <vector>


int max_subset_sum_no_adjacent(std::vector<int> array) {
    if (array.empty()) {
        return 0;
    }

    for (int i = 2; i < array.size(); ++i) {
        array[i] += *std::max_element(array.cbegin(), array.cend() - array.size() + i - 1);
    }

    return *std::max_element(array.cbegin(), array.cend());
}